[vulkan] VulkanTensor, add strides in interface

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39410 [vulkan] Remove unused variables
* #39384 [vulkan] Shaders caching
* #39282 [vulkan] jit passes for vulkan conv2 prepack and fuse with clamp
* #39115 [vulkan] Conv2d with optional clamp
* #39078 [vulkan] addmm support non-vulkan inputs
* **#39077 [vulkan] VulkanTensor, add strides in interface**

We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

Differential Revision: [D21962426](https://our.internmc.facebook.com/intern/diff/D21962426)

IvanKobzarev committed:
[vulkan] VulkanTensor, add strides in interface

[ghstack-poisoned]

IvanKobzarev committed:
drop batch norm stub as fuse_models Conv2-BatchNorm2 works on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

[ghstack-poisoned]

IvanKobzarev committed:
rebase on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.cpp:
```diff
@@ -1025,12 +1025,17 @@ class VulkanTensor::Impl {
         initVulkanContextOnce(), "Vulkan Failed to create Vulkan Context");
     numel_ = std::accumulate(
         std::begin(sizes_), std::end(sizes_), 1, std::multiplies<int64_t>());
+    strides_ = std::vector<int64_t>(sizes_.size());
```
You can consider initializing this in the constructor to avoid a default initialization only to be overwritten later.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOpaqueTensorImpl.h:
```diff
@@ -0,0 +1,48 @@
+// model on Vulkan backend. Strides are not supported on Vulkan side, plan to
+// support them.
+template <typename OpaqueHandle>
+struct CAFFE2_API VulkanOpaqueTensorImpl
```
CAFFE2_API should not be necessary considering that templates are to be both declared and defined in the header file, making the implementation available to downstream users in the header itself.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOpaqueTensorImpl.h:
```diff
@@ -0,0 +1,48 @@
+            device,
+            opaque_handle,
+            sizes) {
+    strides_ = strides.vec();
```
Consider initializing in the constructor's initializer list to avoid an unnecessary default initialization.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOpaqueTensorImpl.h:
```diff
@@ -0,0 +1,48 @@
+      c10::Device device,
+      OpaqueHandle opaque_handle,
+      c10::IntArrayRef sizes,
+      c10::IntArrayRef strides)
```
Suggestion: consider marking parameters and variables as const if they are not expected to change.

IvanKobzarev committed:
Update on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

Differential Revision: [D21962426](https://our.internmc.facebook.com/intern/diff/D21962426)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

Differential Revision: [D21962426](https://our.internmc.facebook.com/intern/diff/D21962426)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

Differential Revision: [D21962426](https://our.internmc.facebook.com/intern/diff/D21962426)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

Differential Revision: [D21962426](https://our.internmc.facebook.com/intern/diff/D21962426)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] VulkanTensor, add strides in interface"


We plan to support strides for Vulkan, but that is not implemented yet.
The main intention of this faking of strides and is_contiguous - to be able to run torchscript mobilenetV2 on Vulkan backend for development and profiling.

This change adds strides to Vulkan interface and overrides strides(), stride(), is_contiguous() of OpaqueTensorImpl for that purpose.

Differential Revision: [D21962426](https://our.internmc.facebook.com/intern/diff/D21962426)

[ghstack-poisoned]
